src/java/org/apache/cassandra/service/accord/CompactTopology.java

```diff
@@ -0,0 +1,341 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service.accord;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.BitSet;
+import java.util.List;
+import java.util.Objects;
+
+import accord.local.Node;
+import accord.topology.Shard;
+import accord.topology.Topology;
+import accord.utils.SortedArrays;
+import accord.utils.TinyEnumSet;
+import org.apache.cassandra.db.TypeSizes;
+import org.apache.cassandra.io.UnversionedSerializer;
+import org.apache.cassandra.io.util.DataInputPlus;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.schema.TableId;
+import org.apache.cassandra.service.accord.api.TokenKey;
+import org.apache.cassandra.service.accord.serializers.TopologySerializers;
+import org.apache.cassandra.utils.BitSetSerializer;
+import org.apache.cassandra.utils.CollectionSerializers;
+import org.apache.cassandra.utils.ImmutableUniqueList;
+
+public class CompactTopology
```

Comment:
This is the core data structure for this patch.  It dedups the following

1) `TableId`
2) `Range` (`TokenRange` without `TableId`)
3) `Node.Id`

In testing different topology shapes, the savings is between 40-70%, and maybe more with higher table/node counts